xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -386,4 +388,60 @@ void reportClientStats(String clusterName, String serverUri, LoadStatsStore load
    */
   void cancelClientStatsReport(String clusterName) {
   }
+
+  /**
+   * An {@link ObjectPool} holding reference and ref-count of an {@link XdsClient} instance.
+   * Once the ref-count gets back to zero, the xds client will be shutdown and later calls of {@lint
+   * #getObject()} will return null.
+   */
+  static final class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
+    private final Object lock = new Object();
+
+    @Nullable
+    private XdsClient xdsClient;
+    private int refCount;
+
+    RefCountedXdsClientObjectPool(XdsClient xdsClient) {
+      this.xdsClient = Preconditions.checkNotNull(xdsClient, "xdsClient");
+    }
+
+    @CheckReturnValue
+    @Nullable
+    @Override
+    public XdsClient getObject() {
+      synchronized (lock) {
+        if (xdsClient != null) {
+          refCount++;
+        }
+        return xdsClient;
+      }
+    }
+
+    @Override
+    public XdsClient returnObject(Object object) {
+      XdsClient xdsClientToShutdown = null;
+
+      synchronized (lock) {
+        Preconditions.checkState(
+            object == xdsClient,
+            "Bug: the returned object '%s' does not match current XdsClient '%s'",
+            object,
+            xdsClient);
+        if (object != null) {
```

Comment:
This check seems strange to me. What is the expectation here? That if getObject() returns null that someone will return the null object?

Reply:
Yes, that is the case `getObject()` returns `null` and later `returnObject(null)`. I'm OK with disallowing `returnObject(null)`. But let's first resolve whether we should allow `getObject()` return null or throw in the other comment.